core/src/main/java/net/kyori/moonshine/MoonshineInvocationHandler.java

```diff
@@ -0,0 +1,143 @@
+    } else if (method.getName().equals("hashCode")
+        && method.getParameterCount() == 0
+        && method.getReturnType() == int.class) {
+      return this.moonshine.hashCode();
+    } else if (method.getName().equals("toString")
+        && method.getParameterCount() == 0
+        && method.getReturnType() == String.class) {
+      return GenericTypeReflector.getTypeName(this.moonshine.proxiedType().getType()) + '@' + this.moonshine.hashCode();
+    }
+
+    // With that out of the way, get rid of nulls in our parameters.
+    // We do not want a null array as that becomes inconvenient to us.
+    if (args == null) {
+      // As an empty array is immutable, there's also no reason not to just cache and reuse it.
+      args = EMPTY_OBJECT_ARRAY;
+    }
+
+    // We have nothing to do if the user has specified a default implementation...
+    if (method.isDefault()) {
+      // ... in which case, find it and invoke it appropriately.
+      final MethodHandle handle = FIND_METHOD_UTIL.findMethod(method, proxy);
+      if (args.length == 0) {
+        return handle.invokeExact(proxy);
+      } else {
+        return handle.invokeWithArguments(args);
+      }
+    }
+
+    // If for some reason the user wants to access the Moonshine instance, we will let them do so,
+    //   though it is not advised.
+    if (method.getReturnType() == Moonshine.class) {
+      return this.moonshine;
+    }
+
+    final @Nullable MoonshineMethod<? extends R> moonshineMethod =
+        this.moonshine.scannedMethods().get(method);
+    if (moonshineMethod == null) {
+      // This is illegal state, and should be reported if encountered.
+      throw new MissingMoonshineMethodMappingException(this.moonshine.proxiedType().getType(), method);
+    }
+
+    final R receiver = moonshineMethod.receiverLocator().locate(method, proxy, args);
+    final I intermediateMessage = this.moonshine.messageSource().messageOf(receiver, moonshineMethod.messageKey());
+    final Map<String, ? extends F> resolvedPlaceholders = this.moonshine.placeholderResolverStrategy().resolvePlaceholders(
+        this.moonshine, receiver, intermediateMessage, moonshineMethod, args);
+    final O renderedMessage = this.moonshine.messageRenderer().render(receiver, intermediateMessage,
+        resolvedPlaceholders, method, this.moonshine.proxiedType().getType());
+
+    if (method.getReturnType() == void.class) {
+      this.moonshine.messageSender().send(receiver, renderedMessage);
+      return null;
+    } else {
+      return renderedMessage;
+    }
+  }
+
+  static {
+    boolean isJava9 = false;
+    try {
+      Optional.class.getDeclaredMethod("ifPresentOrElse", Consumer.class, Runnable.class);
+      isJava9 = true;
+    } catch (final NoSuchMethodException ignored) {
+      // This is not Java 9; nothing to do.
+    }
+    if (isJava9) {
+      FIND_METHOD_UTIL = new Java9FindMethod();
+    } else {
+      FIND_METHOD_UTIL = new Java8FindMethod();
+    }
```

Comment:
imo it would make more sense to create a temporary variable for the value of `FIND_METHOD_UTIL` to prevent the if statement. Something like this: 
 ```
 IFindMethod method = null;
 try {
    // ...
   method = new Java9FindMethod();
 } catch (...) {
   method = new Java8FindMethod();
 }

FIND_METHOD_UTIL = method;
```

Reply:
I originally had this, however this becomes messy once one has to modify for e.g. Java 16, 17, etc. support. I therefore chose to keep it in a way such that we can easily force an infallible assignment at the end which is clear and decisive.